background.js

```diff
@@ -133,79 +90,68 @@ function modifyReferer(e)
 
 
 /*
- * This function is called when the inject-referer.js content script
- * asks for the value to set as document.referrer.
+ * Listener function for storage changes: Forward any changes to the
+ * global variables.
  */
-function handleContentMsg(request, sender, sendResponse) {
-	const conf = findHostConf(request.target, request.referrer);
-	var referrer = null;
-	switch (conf.action)
-	{
-		case "prune":
-			referrer = new URL(request.referrer).origin + "/";
-			break;
-		case "target":
-			referrer = new URL(request.target).origin + "/";
-			break;
-		case "replace":
-			referrer = conf.referer;
-			break;
-		case "remove":
-			referrer = '';
-			break;
-		default:
-			/* "keep" */
-			referrer = request.referrer;
+async function refreshConfig(change, area)
+{
+	let dirty = false;
+	if (area === "sync") {
+		if (change.hasOwnProperty("domains")) {
+			config.domains = change.domains.newValue;
+			dirty = true;
+		}
+		if (change.hasOwnProperty("any")) {
+			config.anyConf = change.any.newValue;
+			dirty = true;
+		}
+		if (change.hasOwnProperty("same")) {
+			config.sameConf = change.same.newValue;
+			dirty = true;
+		}
 	}
-	sendResponse({referrer: referrer});
-}
 
-
-
-/*
- * Escape function from
- * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
- */
-function escapeRegExp(string)
-{
-	return string.replace(/[.*+?^${}()|\[\]\\]/g, '\\$&');
-	// $& means the whole matched string
+	if (dirty) {
+		engine.setConfig(config);
+		await registerContentScript(config);
+	}
 }
 
 
-
 /*
- * Create regular expressions from domains to also match subdomains.
+ * Register (or re-register) our dynamic content script.
  */
-function createDomainRegex()
-{
-	for (let domain of domains)
-	{
-		let pattern = "(\\.|^)" + escapeRegExp(domain.domain) + "$";
-		console.log("domain '" + domain.domain + "', pattern: " + pattern);
-		domain.regexp = new RegExp(pattern);
-	}
-}
+async function registerContentScript(config) {
+	let code = `
+		var engineConfig = JSON.parse('${JSON.stringify(config)}');
+		if (typeof engineInstance === 'object') {
+			engineInstance.setConfig(engineConfig);
+		}
+	`;
 
+	if (registeredContentScript) {
+		registeredContentScript.unregister();
+		registeredContentScript = null;
+	}
 
+	// [CAVEAT]
+	//  We may be run before or after ["engine.js", "content.js"]
+	let newRegisteredContentScript = await browser.contentScripts.register({
+		"matches": ["https://*/*", "http://*/*"],
+		"matchAboutBlank": true,
+		"allFrames": true,
+		"js": [{
+			"code": code
+		}],
+		"runAt": "document_start"
+	});
 
-/*
- * Listener function for storage changes: Forward any changes to the
- * global variables.
- */
-function refreshConfig(change, area)
-{
-	if (area === "sync" && change.hasOwnProperty("domains"))
-	{
-		domains = change.domains.newValue;
-		createDomainRegex();
+	if (registeredContentScript) {
+		registeredContentScript.unregister();
```

Comment:
Why is there another `registeredContentScript.unregister();` call here, after the one in line 133?

Reply:
Hard to explain this, but I will try...

Both `registeredContentScript.unregister` and `browser.contentScripts.register` are asynchronous operations.

The `registerContentScript` is structured like: ...blahblah... | "preemptable" | `await browser.contentScripts.register` | "preemptable" | ...blahblah...
(I am borrowing the "preemptable" term here, it is not really accurate, but should convey the meaning)

What happens is that the first `registeredContentScript.unregister` occurs "concurrently" to `browser.contentScripts.register`.
In the time we `await browser.contentScripts.register(...)`, it is possible that `registerContentScript` is called again, so without an `registeredContentScript.unregister` at the end, we will end up with double registered content scripts, and one of which is orphaned.

Thinking it again, it might be better to `unregister` after `register`, will improve this.

Reply:
Uploaded new commits

Reply:
[According to MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/contentScripts/RegisteredContentScript/unregister) unregister is synchronous, so unregistering the old script only after registering the new one should be fine. Might still lead to weird results if the order of the async register operations resolving could end up different from the order of events causing them.

Reply:
> According to MDN

It did not say "synchronous", did it? Or are you assuming that based on the `Return value -> None.`?
Also, what is important here is the fact that `register` is asynchronous.

BTW, do you know that there is a documentation error there?
Check the return value of `unregister()` in real Firefox and you will see that it is a `Promise`!

Reply:
I just checked and you're right. The documentation doesn't explicitly say anything, and I got used to async function being marked as such. Which leaves the question how to avoid double (or worse) script registrations, in Python I'd just wrap the register/unregister part with an [`asyncio.Lock`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Lock), but I don't see an obvious equivalent in JS.

Reply:
I think that is solved as of https://github.com/airtower-luna/referer-mod/pull/20/commits/386885f5d9f8d3988e72cc9871d5142853fdba43.

Reply:
The double unregister call looks a bit annoying, but yes, it should work even in case some other async call set `registeredContentScript` meanwhile. :slightly_smiling_face:

Reply:
For more details, you might be interested in
https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide/In_depth
https://v8.dev/blog/fast-async#tasks-vs.-microtasks